client/service/src/task_manager/mod.rs

```diff
@@ -294,9 +310,27 @@ impl TaskManager {
 			let mut t1 = self.essential_failed_rx.next().fuse();
 			let mut t2 = self.on_exit.clone().fuse();
 
-			futures::select! {
-				_ = t1 => Err(Error::Other("Essential task failed.".into())),
-				_ = t2 => Ok(()),
+			if self.children.is_empty() {
```

Comment:
This optimization is not required.

Reply:
it is not XD select_all panic if the list is empty https://docs.rs/futures/0.3.5/futures/future/fn.select_all.html

join_all doesn't though :thinking: